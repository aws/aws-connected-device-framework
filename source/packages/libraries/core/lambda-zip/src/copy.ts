import path from 'node:path';
import { promisify } from 'node:util';
import globCb from 'glob';
import { cp, lstat, mkdir } from 'node:fs/promises';

const glob = promisify(globCb);

export type CopyOptions = {
    to: string;
    from: string;
    pattern?: string;
};

async function copyItem(from: string, to: string) {
    await mkdir(path.dirname(to), { recursive: true });
    await cp(from, to, { recursive: true });
}

export async function copy(options: CopyOptions) {
    const to = path.resolve(options.to);
    const from = path.resolve(options.from);
    const pattern = options.pattern;

    const isDirectory = (await lstat(from)).isDirectory();
    if (pattern && !isDirectory)
        throw new Error(`${from} must be a directory to use pattern ${pattern}`);

    // Shortcut to copy just the one requested file
    if (!isDirectory) {
        await copyItem(from, to);
        return;
    }

    // Shortcut to not bother globbing the whole directory
    if (!pattern) {
        await copyItem(from, to);
        return;
    }

    const filesToCopy = await glob(pattern, { cwd: from, dot: true, nodir: true });
    await Promise.all(
        filesToCopy.map((file) => {
            return copyItem(path.resolve(from, file), path.resolve(to, file));
        })
    );
}
