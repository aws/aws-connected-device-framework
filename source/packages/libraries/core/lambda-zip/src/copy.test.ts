import {
    lstatSync,
    mkdirSync,
    mkdtempSync,
    readdirSync,
    readFileSync,
    rmSync,
    writeFileSync,
} from 'fs';
import { tmpdir } from 'os';
import path from 'path';
import { copy } from './copy';

type FileModel = { [key: string]: FileModel | string };
type FileModelEntry = FileModel[''];

function populateFiles(model: FileModelEntry, target = '.') {
    if (typeof model === 'string') {
        writeFileSync(target, model);
        return;
    }

    mkdirSync(target, { recursive: true });
    for (const key in model) populateFiles(model[key], path.join(target, key));
}

function convertHierarchyToModel(target = '.'): FileModelEntry {
    const stat = lstatSync(target);
    if (stat.isFile()) return readFileSync(target, 'utf-8');

    const result: FileModel = {};
    for (const child of readdirSync(target))
        result[child] = convertHierarchyToModel(path.join(target, child));
    return result;
}

const testModel = {
    a: '/a',
    level1a: {
        a: '/level1a/a',
        b: '/level1a/b',
        level2a: {
            a: '/level1a/level2a/a',
            b: '/level1a/level2a/b',
        },
        level2b: {
            a: '/level1a/level2b/a',
        },
    },
    level1b: {
        level2a: {
            level3a: {
                level4a: {
                    a: '/level1b/level2a/level3a/level4a',
                },
            },
        },
    },
};

describe('copy', () => {
    let sourceDir: string;
    let targetDir: string;
    let originalCwd = process.cwd();

    beforeAll(() => {
        sourceDir = mkdtempSync(path.resolve(tmpdir(), 'copy-test-source'));
        process.chdir(sourceDir);

        // Populate the source dir with contents
        populateFiles(testModel);
    });

    afterAll(() => {
        process.chdir(originalCwd);
        rmSync(sourceDir, { force: true, recursive: true });
    });

    beforeEach(() => {
        targetDir = mkdtempSync(path.resolve(tmpdir(), 'copy-test-target'));
    });

    afterEach(() => {
        rmSync(targetDir, { force: true, recursive: true });
    });

    it('copies a single file when targeting file', async () => {
        const to = path.resolve(targetDir, 'a');
        await copy({ to, from: 'a' });
        expect(convertHierarchyToModel(targetDir)).toMatchSnapshot();
    });

    it('copies a whole directory when targeting a directory', async () => {
        const to = path.resolve(targetDir, 'nested');
        await copy({ to, from: path.join(sourceDir, 'level1a', 'level2a') });
        expect(convertHierarchyToModel(targetDir)).toMatchSnapshot();
    });

    it.each(['a', '*', '**/*', 'level1a/**/b'])(
        'copies files matching pattern (%s)',
        async (pattern) => {
            const to = path.resolve(targetDir);
            await copy({ to, from: '', pattern });
            expect(convertHierarchyToModel(targetDir)).toMatchSnapshot();
        }
    );
});
