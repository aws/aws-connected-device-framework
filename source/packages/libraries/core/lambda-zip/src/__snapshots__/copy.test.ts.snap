// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copy copies a single file when targeting file 1`] = `
{
  "a": "/a",
}
`;

exports[`copy copies a whole directory when targeting a directory 1`] = `
{
  "nested": {
    "a": "/level1a/level2a/a",
    "b": "/level1a/level2a/b",
  },
}
`;

exports[`copy copies files matching pattern (*) 1`] = `
{
  "a": "/a",
}
`;

exports[`copy copies files matching pattern (**/*) 1`] = `
{
  "a": "/a",
  "level1a": {
    "a": "/level1a/a",
    "b": "/level1a/b",
    "level2a": {
      "a": "/level1a/level2a/a",
      "b": "/level1a/level2a/b",
    },
    "level2b": {
      "a": "/level1a/level2b/a",
    },
  },
  "level1b": {
    "level2a": {
      "level3a": {
        "level4a": {
          "a": "/level1b/level2a/level3a/level4a",
        },
      },
    },
  },
}
`;

exports[`copy copies files matching pattern (a) 1`] = `
{
  "a": "/a",
}
`;

exports[`copy copies files matching pattern (level1a/**/b) 1`] = `
{
  "level1a": {
    "b": "/level1a/b",
    "level2a": {
      "b": "/level1a/level2a/b",
    },
  },
}
`;
